File: codes/kotlin/010.operator_nullish_coalescing.kt
@Suppress("UNUSED_VARIABLE", "UNCHECKED_CAST")
// to compile and run the code:
// kotlinc operator_nullish_coalescing.kt -include-runtime -d operator_nullish_coalescing.jar && kotlin operator_nullish_coalescing.jar

fun main() {
    println("\n// JavaScript-like Nullish Coalescing Operator (??) in Kotlin")

    fun prettyJsonStringify(anything: Any? = null, indent: String = "    "): String {
        var indentLevel = 0
        fun prettyJsonStringifyInner(anythingInner: Any?, indentInner: String): String {
            if (anythingInner == null) return "undefined"
            if (anythingInner == "null") return "null"
            if (anythingInner == "undefined") return "undefined"
            if (anythingInner is String) return "\"${anythingInner}\""
            if (anythingInner is Number || anythingInner is Boolean) return "${anythingInner}"
            if (anythingInner is MutableList<*>) {
                if (anythingInner.size == 0) {
                    val result = "[]"
                    return result
                }
                indentLevel += 1
                var result = "[\n${indentInner.repeat(indentLevel)}"
                for ((arrayItemIndex, arrayItem) in anythingInner.withIndex()) {
                    result += prettyJsonStringifyInner(arrayItem, indentInner)
                    if ((arrayItemIndex + 1) != anythingInner.size) {
                        result += ",\n${indentInner.repeat(indentLevel)}"
                    }
                }
                indentLevel -= 1
                result += "\n${indentInner.repeat(indentLevel)}]"
                return result
            }
            if (anythingInner is MutableMap<*, *>) {
                if (anythingInner.entries.size == 0) {
                    val result = "{}"
                    return result
                }
                indentLevel += 1
                var result = "{\n${indentInner.repeat(indentLevel)}"
                anythingInner.entries.forEachIndexed { entryIndex, entryItem ->
                    val objectKey = entryItem.key
                    val objectValue = entryItem.value
                    result += "\"${objectKey}\": ${prettyJsonStringifyInner(objectValue, indentInner)}"
                    if ((entryIndex + 1) != anythingInner.entries.size) {
                        result += ",\n${indentInner.repeat(indentLevel)}"
                    }
                }
                indentLevel -= 1
                result += "\n${indentInner.repeat(indentLevel)}}"
                return result
            }
            return "undefined"
        }
        return prettyJsonStringifyInner(anything, indent)
    }

    // in Kotlin, JavaScript-like Nullish Coalescing Operator (??) is Elvis Operator (?:)

    val JSON_OBJECT = mutableMapOf<String, Any?>(
        "foo" to mutableMapOf<String, Any?>(
            "bar" to "baz",
        ),
        "fruits" to mutableListOf<Any?>("apple", "mango", "banana")
    )
    println("JSON_OBJECT: ${prettyJsonStringify(JSON_OBJECT)}")

    println("JSON_OBJECT?.foo?.bar ?? \"not found\": ${try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["bar"] } catch (err: Exception) { null } ?: "not found"}")
    // JSON_OBJECT?.foo?.bar ?? "not found": baz

    println("JSON_OBJECT?.foo?.baz ?? \"not found\": ${try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["baz"] } catch (err: Exception) { null } ?: "not found"}")
    // JSON_OBJECT?.foo?.baz ?? "not found": not found

    println("JSON_OBJECT?.foo?.bar ?? \"not found\": ${prettyJsonStringify(try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["bar"] } catch (err: Exception) { null } ?: "not found")}")
    // JSON_OBJECT?.foo?.bar ?? "not found": "baz"

    println("JSON_OBJECT?.foo?.baz ?? \"not found\": ${prettyJsonStringify(try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["baz"] } catch (err: Exception) { null } ?: "not found")}")
    // JSON_OBJECT?.foo?.baz ?? "not found": "not found"

    println("JSON_OBJECT?.fruits?.[2] ?? \"not found\": ${try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[2] } catch (err: Exception) { null } ?: "not found"}")
    // JSON_OBJECT?.fruits?.[2] ?? "not found": banana

    println("JSON_OBJECT?.fruits?.[5] ?? \"not found\": ${try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[5] } catch (err: Exception) { null } ?: "not found"}")
    // JSON_OBJECT?.fruits?.[5] ?? "not found": not found

    println("JSON_OBJECT?.fruits?.[2] ?? \"not found\": ${prettyJsonStringify(try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[2] } catch (err: Exception) { null } ?: "not found")}")
    // JSON_OBJECT?.fruits?.[2] ?? "not found": "banana"

    println("JSON_OBJECT?.fruits?.[5] ?? \"not found\": ${prettyJsonStringify(try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[5] } catch (err: Exception) { null } ?: "not found")}")
    // JSON_OBJECT?.fruits?.[5] ?? "not found": "not found"
}

File: codes/swift/010.operator_nullish_coalescing.swift
import Foundation

typealias MyObject = [String: Any?]
typealias MyArray = [Any?]

print("\n// JavaScript-like Nullish Coalescing Operator (??) in Swift")

func prettyJsonStringify(_ anything: Any? = nil, indent: String? = "    ") -> String {
    var indentLevel = 0
    func prettyJsonStringifyInnerFunction(_ anything: Any? = nil, _ indent: String? = "    ") -> String {
        guard let anything = anything else {
            return "undefined"
        }
        if (anything as? String == "null") {
            return "null"
        }
        if (anything as? String == "undefined") {
            return "undefined"
        }
        if let anything = anything as? String {
            return "\"\(anything)\""
        }
        if let anything = anything as? NSNumber {
            return "\(anything)"
        }
        if let anything = anything as? Bool {
            return "\(anything)"
        }
        if let anything = anything as? MyArray {
            indentLevel += 1
            var result = "[\n\(String(repeating: indent ?? "    ", count: indentLevel))"
            for (arrayItemIndex, arrayItem) in anything.enumerated() {
                result += prettyJsonStringifyInnerFunction(arrayItem, indent)
                if ((arrayItemIndex + 1) != anything.count) {
                    result += ",\n\(String(repeating: indent ?? "    ", count: indentLevel))"
                }
            }
            indentLevel -= 1
            result += "\n\(String(repeating: indent ?? "    ", count: indentLevel))]"
            return result
        }
        if let anything = anything as? MyObject {
            indentLevel += 1
            var result = "{\n\(String(repeating: indent ?? "    ", count: indentLevel))"
            for (entryIndex, (objectKey, objectValue)) in anything.enumerated() {
                result += "\"\(objectKey)\": \(prettyJsonStringifyInnerFunction(objectValue, indent))"
                if ((entryIndex + 1) != anything.count) {
                    result += ",\n\(String(repeating: indent ?? "    ", count: indentLevel))"
                }
            }
            indentLevel -= 1
            result += "\n\(String(repeating: indent ?? "    ", count: indentLevel))}"
            return result
        }
        return "undefined"
    }
    return prettyJsonStringifyInnerFunction(anything, indent)
}

let JSON_OBJECT: MyObject = [
    "foo": [
        "bar": "baz"
    ] as MyObject,
    "fruits": ["apple", "mango", "banana"] as MyArray
]
print("JSON_OBJECT: \(prettyJsonStringify(JSON_OBJECT))")

if let result = JSON_OBJECT["foo"] as? MyObject {
    if let result = result["bar"] {
        print("JSON_OBJECT?.foo?.bar ?? \"not found\":", result ?? "not found")
    } else {
        print("JSON_OBJECT?.foo?.bar ?? \"not found\": not found")
    }
} else {
    print("JSON_OBJECT?.foo?.bar ?? \"not found\": not found")
}
// JSON_OBJECT?.foo?.bar ?? "not found": baz

if let result = JSON_OBJECT["foo"] as? MyObject {
    if let result = result["baz"] {
        print("JSON_OBJECT?.foo?.baz ?? \"not found\":", result ?? "not found")
    } else {
        print("JSON_OBJECT?.foo?.baz ?? \"not found\": not found")
    }
} else {
    print("JSON_OBJECT?.foo?.baz ?? \"not found\": not found")
}
// JSON_OBJECT?.foo?.baz ?? "not found": not found

print("JSON_OBJECT?.foo?.bar ?? \"not found\":", { () -> Any? in
    guard let result = JSON_OBJECT["foo"] as? MyObject, let result = result["bar"] else {
        return nil
    }
    return result
}() ?? "not found")
// JSON_OBJECT?.foo?.bar ?? "not found": baz

print("JSON_OBJECT?.foo?.baz ?? \"not found\":", { () -> Any? in
    guard let result = JSON_OBJECT["foo"] as? MyObject, let result = result["baz"] else {
        return nil
    }
    return result
}() ?? "not found")
// JSON_OBJECT?.foo?.baz ?? "not found": not found

if let result = JSON_OBJECT["fruits"] as? MyArray {
    let result = result.indices.contains(2) ? result[2] : nil
    print("JSON_OBJECT?.fruits?.[2] ?? \"not found\":", result ?? "not found")
} else {
    print("JSON_OBJECT?.fruits?.[2] ?? \"not found\": not found")
}
// JSON_OBJECT?.fruits?.[2] ?? "not found": banana

if let result = JSON_OBJECT["fruits"] as? MyArray {
    let result = result.indices.contains(5) ? result[5] : nil
    print("JSON_OBJECT?.fruits?.[5] ?? \"not found\":", result ?? "not found")
} else {
    print("JSON_OBJECT?.fruits?.[5] ?? \"not found\": not found")
}
// JSON_OBJECT?.fruits?.[5] ?? "not found": not found

print("JSON_OBJECT?.fruits?.[2] ?? \"not found\":", { () -> Any? in
    guard let result = JSON_OBJECT["fruits"] as? MyArray else {
        return nil
    }
    return result.indices.contains(2) ? result[2] : nil
}() ?? "not found")
// JSON_OBJECT?.fruits?.[2] ?? "not found": banana

print("JSON_OBJECT?.fruits?.[5] ?? \"not found\":", { () -> Any? in
    guard let result = JSON_OBJECT["fruits"] as? MyArray else {
        return nil
    }
    return result.indices.contains(5) ? result[5] : nil
}() ?? "not found")
// JSON_OBJECT?.fruits?.[5] ?? "not found": not found
